src/OrbitGl/GraphTrack.cpp

```diff
@@ -0,0 +1,293 @@
+      target_pos[1] - text_box_size[1] / 2.f);
+
+  Box arrow_text_box(text_box_position, text_box_size, z);
+  Vec3 arrow_extra_point(target_pos[0], target_pos[1], z);
+
+  batcher.AddBox(arrow_text_box, font_color);
+  if (arrow_is_left_directed) {
+    batcher.AddTriangle(
+        Triangle(arrow_text_box.vertices[0], arrow_text_box.vertices[1], arrow_extra_point),
+        font_color);
+  } else {
+    batcher.AddTriangle(
+        Triangle(arrow_text_box.vertices[2], arrow_text_box.vertices[3], arrow_extra_point),
+        font_color);
+  }
+
+  text_renderer.AddText(text.c_str(), text_box_position[0],
+                        text_box_position[1] + text_box_size[1] - layout_->GetTextBoxHeight() +
+                            layout_->GetTextOffset(),
+                        z, text_color, font_size, text_box_size[0]);
+}
+
+template <size_t Size>
+void GraphTrack<Size>::DrawLegend(Batcher& batcher, TextRenderer& text_renderer,
+                                  const std::array<std::string, Size>& series_names,
+                                  const Color& legend_text_color, float z) {
+  uint32_t font_size = layout_->CalculateZoomedFontSize();
+
+  const float kSpaceBetweenLegendSymbolAndText = layout_->GetRightMargin();
+  const float kSpaceBetweenLegendEntries = layout_->GetRightMargin() * 2;
+  float legend_symbol_height = GetLegendHeight() / 2.f;
+  float legend_symbol_width = legend_symbol_height;
+  float x0 = pos_[0] + layout_->GetRightMargin();
+  float y0 = pos_[1] - layout_->GetTrackTabHeight() - layout_->GetTextBoxHeight() / 2.f;
+
+  for (size_t i = 0; i < series_names.size(); ++i) {
+    batcher.AddShadedBox(Vec2(x0, y0 - legend_symbol_height / 2.f),
+                         Vec2(legend_symbol_width, legend_symbol_height), z,
+                         TimeGraph::GetColor(i));
+    x0 += legend_symbol_width + kSpaceBetweenLegendSymbolAndText;
+
+    float legend_text_width = text_renderer.GetStringWidth(series_names[i].c_str(), font_size);
+    Vec2 legend_text_box_size(legend_text_width, layout_->GetTextBoxHeight());
+    Vec2 legend_text_box_position(x0, y0 - layout_->GetTextBoxHeight() / 2.f);
+    text_renderer.AddText(series_names[i].c_str(), legend_text_box_position[0],
+                          legend_text_box_position[1] + layout_->GetTextOffset(), z,
+                          legend_text_color, font_size, legend_text_box_size[0]);
+    x0 += legend_text_width + kSpaceBetweenLegendEntries;
+  }
+}
+
+template <size_t Size>
+void GraphTrack<Size>::DrawSeries(Batcher* batcher, uint64_t min_tick, uint64_t max_tick, float z) {
+  std::array<std::string, Size> series_names = series_.names();
+  std::map<uint64_t, std::array<double, Size>> time_to_series_values = series_.value_map();
+  auto it = time_to_series_values.upper_bound(min_tick);
+  if (it == time_to_series_values.end()) return;
+  if (it != time_to_series_values.begin()) --it;
```

Comment:
I found that both places can use the `GetPreviousOrFirstEntry` method which returns a const_iterator to the previous time series if exists, otherwise returns the first entry.
Thanks for all the suggestions!